### PR TITLE
fix(schematron): handle location attribute not pointing to one node

### PR DESCRIPTION
**Pull request by vincentml**
This pull request fixes a problem that I encountered when updating XSpec Schematron tests from XSpec versions 1.4 and 1.6 to XSpec version 2.1.4. In the newer version of XSpec, the XPath that is provided in the `location` attribute in Schematron tests is evaluated using the `is` operator. The new `is` comparison is more accurate, but this improved accuracy means that XPaths that locate more than one node now fail whereas they did not fail in previous versions of XSpec. The solution, for XSpec users, is to make the XPath location more specific, for example replace `location="sec/p"` with `location="sec[1]/p[2]"`. 

The reason for this pull request is that when the XPath in a location attribute identifies more than one node XSpec terminates abruptly with an error "XPDY0050: A sequence of more than one item is not allowed as the value in 'treat as'". This pull request fixes this problem so that when an XPath identifies more than one node the test fails. The failed test is shown in the report so that the incorrect XPath can be found and corrected more easily. 

Here are log files that show the error message that is encountered before this fix [schematron-027-error.log](https://github.com/xspec/xspec/files/6820185/schematron-027-error.log) and a successful run after this fix [schematron-027-after.log](https://github.com/xspec/xspec/files/6820187/schematron-027-after.log)

### CODE REVIEW

**Comment by AirQuick** [comment]
But, this way, you won't notice erroneous `x:expect-not-report/@location` and `x:expect-not-assert/@location`. For example, if you rewrite `test/schematron-026.xspec` as follows, this pr branch silently succeeds while the current `master` crashes.

```diff
                <x:expect-not-report
-                       location="article/front/article-meta/abstract[1]/sec[@id='sec-title-text-lower-case']" />
+                       location="NONE" />
```

```diff
                <x:expect-not-assert
-                       location="article/front/article-meta/abstract[position() ne 1]/sec[@id='sec-title-text-UPPER-CASE']" />
+                       location="NONE" />
```

So I guess that the predicate in question should return `error((), some friendly message)` instead of returning `false()` when a sequence selected by `@location` is not `instance of node()`.

**Comment by vincentml** [comment]
Hi @AirQuick 

I'm sorry for the delayed reply. Your suggestion to produce an error with a friendly error message is better than false(). I've made this change https://github.com/xspec/xspec/pull/1506/commits/ca6bba6b793091148160bd26b9703bc260bb7faa with a suggestion for a friendly error message that includes the value of `location`.

    XPDY0050: ERROR in location (XPath should point to only one node "contrib/aff")

Your point that `location` is meaningless on `x:expect-not-report` and `x:expect-not-assert` is also a good observation. After coming to the same realization I stopped writing `location` in expect-not-* tests. Perhaps it would be good to update xspec.rnc to formally remove the `location` attribute from `x:expect-not-report` and `x:expect-not-assert`

**Comment by AirQuick** [comment]
After merging the `master` into this pr branch, I removed a test from `schematron-027.xspec` in favor of #1521.

File: src/schematron/schut-to-xspec.xsl [hunk 1]
```diff
@@ -185,7 +185,7 @@
...
 
     <xsl:template match="@location" as="text()" mode="make-predicate">
-        <xsl:text expand-text="yes">[(${x:known-UQName('x:context')}/root()/({.}) treat as node()) is {x:known-UQName('x:select-node')}(${x:known-UQName('x:context')}/root(), @location, preceding-sibling::{x:known-UQName('svrl:ns-prefix-in-attribute-values')}, {parent::element() => x:xslt-version()})]</xsl:text>
+        <xsl:text expand-text="yes">[if (not((${x:known-UQName('x:context')}/root()/({.})) instance of node() and ({x:known-UQName('x:select-node')}(${x:known-UQName('x:context')}/root(), @location, preceding-sibling::{x:known-UQName('svrl:ns-prefix-in-attribute-values')}, {parent::element() => x:xslt-version()})) instance of node())) then error(QName('http://www.w3.org/2005/xqt-errors', 'XPDY0050'), 'ERROR in location (XPath should point to only one node "{.}")') else (${x:known-UQName('x:context')}/root()/({.}) treat as node()) is {x:known-UQName('x:select-node')}(${x:known-UQName('x:context')}/root(), @location, preceding-sibling::{x:known-UQName('svrl:ns-prefix-in-attribute-values')}, {parent::element() => x:xslt-version()})]</xsl:text>
```
Comment (AirQuick) [inline src/schematron/schut-to-xspec.xsl:188]: Putting a user-provided string into `''` (the 2nd parameter of `error()`) without escaping its `'` is fragile. These tests are now broken:
```console
$ bin/xspec.sh -s test/issue-396.xspec
...
Running Tests...
Syntax error near {...t()[. eq 'above-mentioned']...} at char 703 in expression in xsl:sequence/@select on line 250 column 46 of issue-396-compiled.xsl:
  XPST0003: expected ")", found name "above-mentioned"
Syntax error near {...t()[. eq 'above-mentioned']...} at char 703 in expression in xsl:sequence/@select on line 256 column 43 of issue-396-compiled.xsl:
  XPST0003: expected ")", found name "above-mentioned"
Errors were reported during stylesheet compilation

*** Error running the test suite
$ bin/xspec.sh -s test/schematron-026.xspec 
...
Running Tests...
Syntax error near {...sec-title-text-lower-case']...} at char 676 in expression in xsl:sequence/@select on line 215 column 46 of schematron-026-compiled.xsl:
  XPST0003: expected ")", found name "sec-title-text-lower-case"
...
*** Error running the test suite
$ bin/xspec.sh -s test/xspec-sch.xspec
...
Running Tests...
Syntax error near {...ith(@label, 'In function-pa...} at char 903 in expression in xsl:sequence/@select on line 225 column 46 of xspec-sch-compiled.xsl:
  XPST0003: expected ")", found name "In"
...
*** Error running the test suite
```

I'm fixing it.

Comment (AirQuick) [reply src/schematron/schut-to-xspec.xsl]: Fixed in e7084cf47009e230429e51fba59be3b2cf5367b8.

**Comment by AirQuick** [comment]
@vincentml 
While working on https://github.com/xspec/xspec/pull/1506#discussion_r684690041, I found some other issues and fixed them as mentioned in the commit message of e7084cf47009e230429e51fba59be3b2cf5367b8. Does it look fine?

> Your point that `location` is meaningless on `x:expect-not-report` and `x:expect-not-assert` is also a good observation. After coming to the same realization I stopped writing `location` in expect-not-* tests. Perhaps it would be good to update xspec.rnc to formally remove the `location` attribute from `x:expect-not-report` and `x:expect-not-assert`

I didn't mean to go that far. `@location` in `x:expect-not-report` and `x:expect-not-assert` seems still useful when the test author really wants to make sure `report` or `assert` is not thrown in a specific location, like this:

https://github.com/xspec/xspec/blob/9ade2b95c7ebb3de7da736b2d757e907042df3c7/test/schematron-012.xspec#L34-L40

**Comment by AirQuick** [comment]
@vincentml 
Another option would be to allow `x:expect-*/@location` to specify multiple nodes. More specifically:

A: One or more nodes specified by `x:expect-*/@location`
B: One node specified by SVRL `@location`
Predicate: `[A intersect B]`

**Comment by vincentml** [comment]
@AirQuick 

Thank you for working on this and greatly improving upon my pull request. The code is now easier to read and handles more similar cases.

Maybe removing `@location` from `x:expect-not-assert` and `x:expect-not-report` is not the best idea then if `@location` is useful in those contexts. There is a risk of false-positive test outcomes when using `@location` on `x:expect-not-assert` and `x:expect-not-report`. When using `x:expect-not-assert` or `x:expect-not-report` with `@location` the assert may fail, or the report may succeed, at a different XPath location, even slightly different, than the one described in `@location` resulting in a problem not being detected by XSpec because of the `@location`. This can give the user a false sense of security by seeing their tests passing when really the expectations in their test scenario are not detecting the problem. Perhaps this risk is just something that the user needs to have an awareness of. After all, the onus is always on the user to ensure that their test scenarios are written correctly to detect the problems that they want to know about.

**Comment by vincentml** [comment]
@AirQuick 

Since each `svrl:failed-assert` and `svrl:successful-report` occur with at one specific location indicated by their `@location` attribute it is precise to match one node using `is`. That is an interesting idea to allow multiple nodes using `intersect`. Perhaps, for example, a user would write a scenario that contains something like `@location="//p"` to test that an assert (or report) does (or does not) occur on some element `p`. I can see how that could help with the problem of `@location` on `x:expect-not-*` potentially creating false-positive test results by creating more flexibility. It gives the user the choice to write XPath in `@location` that selects one node or selects multiple nodes. The `svrl:*/@location` generated by Schematron should point to one  node, so the the comparison would check that one or more nodes selected by `x:expect-*/@location` intersects one node selected by `svrl:*/@location`. I think this is would be a good improvement!

**Comment by AirQuick** [comment]
> There is a risk of false-positive test outcomes when using `@location` on `x:expect-not-assert` and `x:expect-not-report`.

Good point.
I guess the biggest risk is unintentionally writing an XPath which points to an empty sequence. Fortunately, the risk has been reduced greatly on XSpec v2.x, and with this PR, more friendly error messages.

**Comment by AirQuick** [comment]
@vincentml 

Thanks for the insight. I'm merging this, keeping the current `is` operator.

As you pointed out, I see some use cases in using `intersect` especially on `x:expect-not-*`. At some point in future, we may want to consider introducing it. Right now, I'm not sure if it should be applied to all `@location` or only `x:expect-not-*/@location` or a new attribute (`@some-locations` for example).

**Comment by AirQuick** [comment]
[Updated](https://github.com/xspec/xspec/wiki/Testing-Schematron-with-Text-Nodes/_compare/6105c3dfc8a865a35aa334edf662d6fbfeb5c72f...8998beacc9b7e8670ae69da6ad3b122d3611b51a) Wiki [_Testing Schematron with Text Nodes_](https://github.com/xspec/xspec/wiki/Testing-Schematron-with-Text-Nodes#setting-global-parameters-for-schematron-to-visit-text-nodes) page to reflect this change.

**Pull request merged**